kimchi/src/circuits/polynomials/foreign_mul/gate.rs

```diff
@@ -0,0 +1,505 @@
+//! Foreign field multiplication gate
+
+use std::collections::HashMap;
+
+use ark_ff::{FftField, PrimeField, SquareRootField, Zero};
+use ark_poly::{univariate::DensePolynomial, Evaluations, Radix2EvaluationDomain as D};
+use array_init::array_init;
+use num_bigint::BigUint;
+use rand::{prelude::StdRng, SeedableRng};
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+
+use crate::{
+    alphas::Alphas,
+    circuits::{
+        argument::{Argument, ArgumentType},
+        constraints::ConstraintSystem,
+        domains::EvaluationDomains,
+        expr::{self, l0_1, Environment, E},
+        gate::{CircuitGate, GateType},
+        polynomial::COLUMNS,
+        wires::GateWires,
+    },
+};
+
+use super::{ForeignMul0, ForeignMul1};
+
+impl<F: FftField + SquareRootField> CircuitGate<F> {
+    /// Create foreign multiplication gate
+    pub fn create_foreign_mul(wires: &[GateWires; 8]) -> Vec<Self> {
+        vec![
+            /* Input: a */
+            CircuitGate {
+                typ: GateType::ForeignMul0,
+                wires: wires[0],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul0,
+                wires: wires[1],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul1,
+                wires: wires[2],
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::ForeignMul2,
```

Comment:
Since `ForeignMul2` is only useful to apply the lookups in the row after `ForeignMul1`, it probably makes more sense to use `ForeignMul1, Next` as the trigger for the lookups vs `ForeignMul2, Curr`.

This can be done in the follow-up lookup PR, doesn't need to be here.

Reply:
Thanks - noted!